File: Matlab-code/testModel.m
function [targetPredictions,correlation,MAE,RMSE] = testModel(testImages,testBasisFun,testTarget,V,beta,W,trainBasisFunMean,trainBasisFunStd)


    showFigures = 1;

    targetTrainMean = trainBasisFunMean(1);
    targetTrainStd = trainBasisFunStd(1);
    nLatentVars = size(V,2);
    nVoxels = size(V,1);
    nTest = size(testImages,1);
    
    wStar = W(:,1);
    
    Delta=spdiags(beta,0,nVoxels,nVoxels);
    SigmaZinv=eye(nLatentVars)+V'*Delta*V;
    SigmaZ=inv(SigmaZinv);
    
    
    
    DeltaTimesWstar = Delta*wStar;
    aux = V'*DeltaTimesWstar;
    %posteriorVarTarget=1/(1/priorVarTarget+wStar'*DeltaTimesWstar-aux'*SigmaZ*aux);
    posteriorVarTarget=1/(wStar'*DeltaTimesWstar-aux'*SigmaZ*aux);

    gap = testImages'-W(:,2:end)*testBasisFun(:,2:end)';
    aux2 = V'*Delta*gap;
    standTargetPredictions = posteriorVarTarget.*(DeltaTimesWstar'*gap-aux'*SigmaZ*aux2)';
   

    %take predictions back to the original target space
    targetPredictions = standTargetPredictions.*targetTrainStd+targetTrainMean;
    
    correlation=corr(testTarget,targetPredictions);
    RMSE=sqrt(sum((testTarget-targetPredictions).^2)/nTest);
    MAE=mean(abs(testTarget-targetPredictions));

    fprintf('test correlation : %.4f\n', correlation)
    fprintf('test RMSE : %.4f\n', RMSE)
    fprintf('test MAE : %.4f\n', MAE)
    fprintf('\n')

    if showFigures
        figure, scatter(testTarget,targetPredictions)
        hold on
        plot(testTarget,testTarget)
        xlabel('true score')
        ylabel('predicited score')
        title('Predictions')
        hold off
    end
   
    return

File: python-code/testModel.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""

"""


import numpy as np
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error

def testModel(testImages,testBasisFun,testTarget,V,beta,W,trainBasisFunMean,trainBasisFunStd):

    targetTrainMean = trainBasisFunMean[0]
    targetTrainStd = trainBasisFunStd[0]
    nLatentVars = np.shape(V)[1]
    nVoxels = np.shape(V)[0]
    
    w_star = np.reshape(W[:,0],(nVoxels,1))
    
    
    SigmaZ_inv=np.eye(nLatentVars)+np.dot(np.transpose(V),beta*V)
    SigmaZ=np.linalg.inv(SigmaZ_inv)
   
    
    DeltaTimesWstar = beta*w_star
    aux=np.dot(np.transpose(V),beta*w_star)
    posteriorVarTarget=1/(np.dot(np.transpose(w_star),beta*w_star)-np.linalg.multi_dot((np.transpose(aux),SigmaZ,aux)))
    
    
   
    gap=np.transpose(testImages)-np.dot(W[:,1:],np.transpose(testBasisFun[:,1:]))
    aux2=np.dot(np.transpose(V),beta*gap)
    standTargetPredictions=posteriorVarTarget*np.transpose((np.dot(np.transpose(beta*w_star),gap)-np.linalg.multi_dot((np.transpose(aux),SigmaZ,aux2))))
    
    
    #take predictions back to the original target space
    targetPredictions=standTargetPredictions*targetTrainStd+targetTrainMean
    
    #compute correlation and error
    correlation=np.corrcoef(np.transpose(testTarget),np.transpose(targetPredictions))[0,1]
    RMSE=np.sqrt(mean_squared_error(testTarget,targetPredictions))
    MAE=mean_absolute_error(testTarget,targetPredictions)
   
    

    print('correlation : ', correlation, '\n')
    print('RMSE : ', RMSE, '\n')
    print('MAE : ', MAE, '\n')
    print('\n')

    
    
    return targetPredictions,correlation,MAE,RMSE